docs/proposals/upstream-settings.md

```diff
@@ -0,0 +1,210 @@
+)
+
+type UpstreamSettingsPolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of the UpstreamSettingsPolicy.
+    Spec UpstreamSettingsPolicySpec `json:"spec"`
+
+    // Status defines the state of the UpstreamSettingsPolicy.
+    Status gatewayv1alpha2.PolicyStatus `json:"status,omitempty"`
+}
+
+type UpstreamSettingsSpec struct {
+    // TargetRefs identifies API object(s) to apply the policy to.
+    // Objects must be in the same namespace as the policy.
+    // Support: Service
+    TargetRefs []gatewayv1alpha2.LocalPolicyTargetReference `json:"targetRefs"`
+
+    // ZoneSize is the size of the shared memory zone used by the upstream. This memory zone is used to share
+    // the upstream configuration between nginx worker processes. The more servers that an upstream has,
+    // the larger memory zone is required.
+    // +optional
+    ZoneSize *Size `json:"zoneSize,omitempty"`
+
+    // KeepAlive defines the keep-alive settings.
+    // +optional
+    KeepAlive *UpstreamKeepAlive `json:"keepAlive,omitempty"`
+}
+
+// UpstreamKeepAlive defines the keep-alive settings for upstreams.
+type UpstreamKeepAlive struct {
+    // Connections sets the maximum number of idle keepalive connections to upstream servers that are preserved
+    // in the cache of each nginx worker process. When this number is exceeded, the least recently used
+    // connections are closed.
+    // +optional
+    Connections *int32 `json"connections,omitempty"`
+
+    // Requests sets the maximum number of requests that can be served through one keep-alive connection.
+    // After the maximum number of requests are made, the connection is closed.
+    // +optional
+    Requests *int32 `json:"requests,omitempty"`
+
+    // Time defines the maximum time during which requests can be processed through one keep-alive connection.
+    // After this time is reached, the connection is closed following the subsequent request processing.
+    // +optional
+    Time *Duration `json:"time,omitempty"`
+
+    // Timeout defines the keep-alive timeout for upstreams.
+    // +optional
+    Timeout *Duration `json:"timeout,omitempty"`
+}
+
+// Duration is a string value representing a duration in time.
+// The format is a subset of the syntax parsed by Golang time.ParseDuration.
+// Examples: 1h, 12m, 30s, 150ms.
+type Duration string
+
+// Size is a string value representing a size. Size can be specified in bytes, kilobytes (suffix k),
+// or megabytes (suffix m).
+// Examples: 1024, 8k, 1m.
+type Size string
+```
+
+Since this Policy only applies to `http` upstreams, there's no way to set the zone size for `stream` upstreams. For now, we can introduce a global `zoneSize` variable in the `NginxProxy` resource that will set the zone size for all upstreams. Then this Policy would override that global setting on upstreams that it attaches to.
```

Comment:
Can we make this its own section so we can easily find/link to it in the future?